packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,232 @@
+
+        const taskManagerListener = taskManager.onDelete(id => {
+            const toDispose = this.taskToDispose.get(id);
+            if (toDispose !== undefined) {
+                toDispose.dispose();
+                this.taskToDispose.delete(id);
+            }
+        });
+        this.toDispose.push(taskManagerListener);
+        this.taskConfigurations.init(this);
+    }
+
+    async setWorkspaceRoot(workspaceRoot: string): Promise<void> {
+        this.workspaceRoot = workspaceRoot;
+        await this.taskConfigurations.setWorkspaceRoot(workspaceRoot);
+        return Promise.resolve();
+    }
+
+    dispose() {
+        this.client = undefined;
+        this.toDispose.dispose();
+    }
+
+    getTasks(): Promise<string[]> {
+        return Promise.resolve(this.taskConfigurations.getTaskLabels());
+    }
+
+    run(task: ITaskOptions | string): Promise<ITaskInfo> {
+        if (typeof task === 'string') {
+            return new Promise((resolve, reject) => {
+                this.taskConfigurations.getTask(task)
+                    .then(t => this.doRun(t).then(taskInfo => resolve(taskInfo)))
+                    .catch(e => {
+                        reject(new Error(e));
+                    });
+            });
+
+        } else {
+            return this.doRun(task);
+        }
+    }
+
+    protected doRun(options: ITaskOptions): Promise<ITaskInfo> {
+        let task: Task;
+        let proc: TerminalProcess | RawProcess;
+        const processOptions = (isWindows && options.windowsProcessOptions !== undefined) ?
+            options.windowsProcessOptions : options.processOptions;
+
+        const command = processOptions.command;
+
+        let cwd;
+        if (processOptions.options === undefined) {
+            processOptions.options = {
+                'cwd': '',
+                'env': ''
+            };
+        }
+
+        cwd = options.cwd;
+        if (cwd !== undefined) {
+            cwd = cwd.replace(/\$workspace/gi, FileUri.fsPath(this.workspaceRoot));
+        }
+
+        // Use task's cwd wit hspawned process
+        processOptions.options.cwd = cwd;
+        // pass node env to new process
+        processOptions.options.env = process.env;
+
+        try {
+            // check that command exists before creating process for it
+            if (this.findCommand(command, cwd) === undefined) {
+                throw (`Command not found: ${command}`);
+            }
+
+            // use terminal or raw process
+            if (options.processType === 'terminal') {
+                this.logger.info('Task: creating underlying terminal process');
+                proc = this.terminalProcessFactory(<TerminalProcessOptions>processOptions);
+            } else {
+                this.logger.info('Task: creating underlying raw process');
+                proc = this.rawProcessFactory(<RawProcessOptions>processOptions);
+            }
+
+            task = new Task(
+                this.taskManager,
```

Comment:
I'm surprised tasks knows about it's manager this way ?
Should be via inversify and we use a TaskFactory

Reply:
right - done